File: sdk/python/src/roder_sdk/hosted.py
"""Hosted multi-tenant connection helpers (roadmap phase 72, Task 6).

Hosted Roder authenticates at the WebSocket handshake with a bearer
credential in the ``Authorization`` header — the gateway always rejects
query-string credentials. :class:`HostedClient` wraps the standard RPC
client with typed helpers for ``hosted/*`` methods; raw JSON-RPC access
stays available via ``client.raw_request`` for forward-compatible hosted
methods.

Token refresh/reconnect: connections authenticate once at handshake time,
so refreshing a token means reconnecting. :meth:`HostedClient.reconnect`
builds a fresh transport using the token provider. Requests in flight when
a connection drops fail with a transport error and are NEVER replayed
automatically — callers retry mutating requests themselves because only
they know whether the operation is idempotent.
"""

from __future__ import annotations

import inspect
from collections.abc import Awaitable, Callable
from typing import Any

from .client import RoderRpcClient
from .transports import WebSocketTransport

TokenProvider = Callable[[], "str | Awaitable[str]"]


class HostedClient:
    def __init__(self, options: dict[str, Any], client: RoderRpcClient) -> None:
        self._options = options
        self.client = client

    @classmethod
    async def connect(
        cls,
        url: str,
        *,
        token: str | None = None,
        token_provider: TokenProvider | None = None,
        headers: dict[str, str] | None = None,
        connector: Callable[..., Awaitable[Any]] | None = None,
    ) -> "HostedClient":
        """Connects and authenticates against a hosted Roder gateway."""
        options = {
            "url": url,
            "token": token,
            "token_provider": token_provider,
            "headers": headers,
            "connector": connector,
        }
        client = await _hosted_rpc_client(options)
        return cls(options, client)

    async def reconnect(self) -> None:
        """Re-authenticates with a fresh credential and replaces the
        connection. In-flight requests on the old connection fail; nothing
        is replayed."""
        next_client = await _hosted_rpc_client(self._options)
        previous = self.client
        self.client = next_client
        await previous.close()

    async def whoami(self) -> dict[str, Any]:
        return await self.client.call("hosted/whoami", {})

    async def create_service_account(self, display_name: str) -> dict[str, Any]:
        return await self.client.call(
            "hosted/service_accounts/create", {"displayName": display_name}
        )

    async def revoke_service_account(self, key_id: str) -> dict[str, Any]:
        return await self.client.call("hosted/service_accounts/revoke", {"keyId": key_id})

    async def list_hooks(self) -> dict[str, Any]:
        return await self.client.call("hosted/hooks/list", {})

    async def create_hook(self, hook: dict[str, Any]) -> dict[str, Any]:
        return await self.client.call("hosted/hooks/create", {"hook": hook})

    async def delete_hook(self, hook_id: str) -> dict[str, Any]:
        return await self.client.call("hosted/hooks/delete", {"hookId": hook_id})

    async def audit_list(self) -> dict[str, Any]:
        return await self.client.call("hosted/audit/list", {})

    def notifications(self):
        return self.client.notifications()

    async def close(self) -> None:
        await self.client.close()


async def _resolve_token(options: dict[str, Any]) -> str | None:
    provider: TokenProvider | None = options.get("token_provider")
    if provider is not None:
        token = provider()
        if inspect.isawaitable(token):
            token = await token
        return str(token)
    return options.get("token")


async def _hosted_rpc_client(options: dict[str, Any]) -> RoderRpcClient:
    token = await _resolve_token(options)
    headers: dict[str, str] | None = options.get("headers")
    has_external_auth = bool(headers) and any(
        key.lower() == "authorization" for key in headers
    )
    if not token and not has_external_auth:
        raise ValueError(
            "hosted connections require a token, token_provider, or an Authorization header"
        )
    transport = await WebSocketTransport.connect(
        options["url"],
        token=token,
        headers=headers,
        connector=options.get("connector"),
    )
    return RoderRpcClient(transport)

File: sdk/typescript/src/hosted.ts
// Hosted multi-tenant connection helpers (roadmap phase 72, Task 6).
//
// Hosted Roder authenticates at the WebSocket handshake with a bearer
// credential (static key or `rk_sa_*` service-account key) in the
// `Authorization` header — query-string credentials are always rejected by
// the gateway. `HostedClient` wraps the standard RPC client with typed
// helpers for `hosted/*` methods; raw JSON-RPC access stays available for
// forward-compatible hosted methods via `client.rawRequest`.
//
// Token refresh/reconnect: connections are authenticated once at handshake
// time, so refreshing a token means reconnecting. `reconnect()` builds a
// fresh transport using the token provider. Requests that were in flight
// when a connection dropped fail with a transport error and are NEVER
// replayed automatically — mutating requests must be retried by the
// caller, who knows whether the operation is idempotent.

import { RoderRpcClient } from "./client.js";
import { WebSocketTransport, type WebSocketFactory } from "./transports.js";

export interface HostedConnectOptions {
  /** Gateway URL, e.g. `wss://roder.example.com`. */
  url: string;
  /** Static bearer token or service-account key. */
  token?: string;
  /**
   * Called whenever a (re)connection needs a credential; takes precedence
   * over `token`. Use this for short-lived externally-issued tokens.
   */
  tokenProvider?: () => string | Promise<string>;
  /** Extra auth headers supplied by an external auth layer. */
  headers?: Record<string, string>;
  protocols?: string[];
  webSocketFactory?: WebSocketFactory;
}

export interface HostedWhoami {
  tenant: { tenantId: string; displayName?: string };
  principal: Record<string, unknown>;
  role: string;
  scopes: string[];
}

export interface HostedServiceAccountKey {
  keyId: string;
  /** Returned exactly once; the gateway stores only a hash. */
  token: string;
}

export interface HostedHookDefinition {
  id: string;
  scope: "tenant" | "system";
  eventKinds: string[];
  url: string;
  /** `env:NAME` reference; raw secrets are rejected by the gateway. */
  signingSecretRef?: string;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export class HostedClient {
  private constructor(
    private readonly options: HostedConnectOptions,
    public client: RoderRpcClient,
  ) {}

  /** Connects and authenticates against a hosted Roder gateway. */
  static async connect(options: HostedConnectOptions): Promise<HostedClient> {
    const client = await hostedRpcClient(options);
    return new HostedClient(options, client);
  }

  /**
   * Re-authenticates with a fresh credential from the token provider and
   * replaces the underlying connection. In-flight requests on the old
   * connection fail; nothing is replayed.
   */
  async reconnect(): Promise<void> {
    const next = await hostedRpcClient(this.options);
    const previous = this.client;
    this.client = next;
    await previous.close();
  }

  whoami(): Promise<HostedWhoami> {
    return this.client.call("hosted/whoami", {});
  }

  createServiceAccount(displayName: string): Promise<HostedServiceAccountKey> {
    return this.client.call("hosted/service_accounts/create", { displayName });
  }

  revokeServiceAccount(keyId: string): Promise<{ revoked: boolean }> {
    return this.client.call("hosted/service_accounts/revoke", { keyId });
  }

  listHooks(): Promise<{ hooks: HostedHookDefinition[] }> {
    return this.client.call("hosted/hooks/list", {});
  }

  createHook(hook: HostedHookDefinition): Promise<{ hook: HostedHookDefinition }> {
    return this.client.call("hosted/hooks/create", { hook });
  }

  deleteHook(hookId: string): Promise<{ deleted: boolean }> {
    return this.client.call("hosted/hooks/delete", { hookId });
  }

  auditList(): Promise<{ records: Record<string, unknown>[] }> {
    return this.client.call("hosted/audit/list", {});
  }

  notifications() {
    return this.client.notifications();
  }

  close(): Promise<void> | void {
    return this.client.close();
  }
}

async function hostedRpcClient(options: HostedConnectOptions): Promise<RoderRpcClient> {
  const token = options.tokenProvider ? await options.tokenProvider() : options.token;
  if (!token && !options.headers?.Authorization && !options.headers?.authorization) {
    throw new Error(
      "hosted connections require a token, tokenProvider, or an Authorization header",
    );
  }
  const transport = new WebSocketTransport({
    url: options.url,
    token,
    protocols: options.protocols,
    webSocketFactory: options.webSocketFactory,
    headers: options.headers,
  });
  return new RoderRpcClient(transport);
}
